tt_metal/host_api.hpp

```diff
@@ -252,10 +264,37 @@ void SetRuntimeArgs(const Program &program, KernelHandle kernel, const std::vari
  * | kernel_id    | ID of the kernel that will receive the runtime args                    | KernelHandle (uint64_t)                                |                                                                            | Yes      |
  * | core_spec    | Location of Tensix core(s) where the runtime args will be written      | const std::vector<CoreCoord> &                         | Any set of logical Tensix core coordinates on which the kernel is placed   | Yes      |
- * | runtime_args | The runtime args to be written                                         | const std::vector< vector<uint32_t> > &                | outer vector size must be equal to size of core_spec vector                | Yes      |
+ * | runtime_args | The runtime args to be written                                         | const std::vector< vector<uint32_t> > &                | Outer vector size must be equal to size of core_spec vector                | Yes      |
  */
 void SetRuntimeArgs(const Program &program, KernelHandle kernel, const std::vector< CoreCoord > & core_spec, const std::vector< std::vector<uint32_t> > &runtime_args);
 
+/**
+ * Set runtime args for a kernel that are sent to the specified cores using the command queue. This API must be used when Asynchronous Command Queue Mode is enabled.
+ *
+ * Return value: void
+ *
+ * | Argument     | Description                                                            | Type                                                   | Valid Range                                                                | Required |
+ * |--------------|------------------------------------------------------------------------|--------------------------------------------------------|----------------------------------------------------------------------------|----------|
+ * | device       | The device that runtime args are being written to.                     | Device*                                                |                                                                            | Yes      |
+ * | kernel       | The kernel that will recieve these runtime args.                       | std::shared_ptr<Kernel>                                |                                                                            | Yes      |
+ * | core_spec    | Location of Tensix core(s) where the runtime args will be written      | const std::variant<CoreCoord,CoreRange,CoreRangeSet> & | Any set of logical Tensix core coordinates on which the kernel is placed   | Yes      |
+ * | runtime_args | The runtime args to be written                                         | std::shared_ptr<RuntimeArgs>                           |                                                                            | Yes      |
+*/
+void SetRuntimeArgs(Device* device, const std::shared_ptr<Kernel> kernel, const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec, std::shared_ptr<RuntimeArgs> runtime_args);
```

Comment:
kernel apis accept `KernelHandle`, can we update these to be consistent with that?

Reply:
Hey @abhullar-tt, as discussed offline passing in a `KernelHandle` requires us to also pass in a ptr/ref to the program object, which can lead to memory corruption `issues` in async mode.

Documenting an example here to illustrate the failure mode:
- `interleaved_to_sharded_multi_core` in `sharded_op_multi_core.cpp` creates a program object locally.
- `SetRuntimeArgs(program, kernel_id, ...)` is asynchronously called inside this function.
- Main thread proceeds to the end of this function and calls:
- `return {.program = std::move(program) ...`
- Program given to worker thread has now been moved/invalidated

As discussed, long term we should be using `shared_ptr` to pass around programs as well. I'll track this in a separate issue.